### PR TITLE
Setup refactor

### PR DESCRIPTION
**Pull request by mindw**
Attempt to modernize setup.py:

* use setuptools requirements for handle packages and versions.
* use setuptools_scm for git based version management.
* rely on setuptools cython support
* some trivial refactoring.

For your consideration.

### CODE REVIEW

File: RELEASING.rst [hunk 1]
```diff
@@ -39,23 +56,15 @@ Updating the online documentation site
...
-  $ rm -rf _build/html
-  $ make html
+  $ python setup.py build_sphinx
 
 * Make a backup and upload the files in the doc site (xodo)::
 
   $ export UPSTREAM="/home/blosc/srv/www/bcolz.blosc.org"
   $ ssh blosc@xodo.blosc.org "mv $UPSTREAM/docs/html $UPSTREAM/docs/html.bck"
-  $ scp -r _build/html blosc@xodo.blosc.org:$UPSTREAM/docs
+  $ scp -r build/sphinx/html blosc@xodo.blosc.org:$UPSTREAM/docs
```
Comment (esc) [inline RELEASING.rst]: yeah, this is actually all not true anymore. I actually do a very different and custom dance to deploy the docs due to the way in which I use xodo and the way the docs are deployed. I'll fix the instructions at some later point, so maybe you could just say: "WARNING: these instructions are out of date" here.

File: setup.py [hunk 2]
```diff
@@ -205,10 +101,21 @@ def check_import(pkgname, pkgver):
 Operating System :: Microsoft :: Windows
 Operating System :: Unix
 """
-setup(name="bcolz",
-      version=VERSION,
-      description='columnar and compressed data containers.',
-      long_description="""\
+
+tests_require = ['mock', 'tox']
```
Comment (esc) [inline setup.py]: don't thing we have a tox config?

File: setup.py [hunk 3]
```diff
@@ -205,10 +101,21 @@ def check_import(pkgname, pkgver):
...
+
+if v < (3,):
+    tests_require.append('unittest2')
+    
+setup(
+    name="bcolz",
+    use_scm_version={
```
Comment (esc) [inline setup.py]: Haven't seen this before, could you say how it works.

File: setup.py [hunk 4]
```diff
@@ -205,10 +101,21 @@ def check_import(pkgname, pkgver):
...
 """
-setup(name="bcolz",
-      version=VERSION,
-      description='columnar and compressed data containers.',
-      long_description="""\
+
+tests_require = ['mock', 'tox']
+
+if v < (3,):
+    tests_require.append('unittest2')
```
Comment (esc) [inline setup.py]: `unittest2` only needed on Py 2.6?

File: setup.py [hunk 5]
```diff
@@ -219,29 +126,45 @@ def check_import(pkgname, pkgver):
...
+    ],
+    cmdclass={'build_ext': BuildExtNumpyInc},
+    install_requires=['numpy>=1.7'],
+    setup_requires=[
+        'cython>=0.22', 
+        'numpy>=1.7', 
+        'setuptools-scm>1.5.4'
+    ],
+    tests_require=tests_require,
+    extras_require=dict(
```
Comment (esc) [inline setup.py]: Now that we have extras could you update the installation instructions to say something about:

```
pip install bcolz[extras]
```

And maybe also mention that bcolz releases can be installed via conda through the standard continuum channels.
(It usually takes a while for the to be built, experience shows, but usually not longer than a day)
```
conda install bcolz
```

File: setup.py [hunk 6]
```diff
@@ -219,29 +126,45 @@ def check_import(pkgname, pkgver):
...
+            library_dirs=lib_dirs,
+            libraries=libs,
+            extra_link_args=LFLAGS,
+            extra_compile_args=CFLAGS
+        )
+    ],
+    cmdclass={'build_ext': BuildExtNumpyInc},
+    install_requires=['numpy>=1.7'],
+    setup_requires=[
+        'cython>=0.22',
```
Comment (esc) [inline setup.py]: I am not sure about this dependency, but see also:

https://github.com/Blosc/bcolz/issues/220

**Comment by esc** [comment]
Can you also change the releasing instructions to list ``twine`` as upload tool. It seems to be the recommended way to go about things nowadays.

File: setup.py (hunk 2 above)
Comment (mindw) [inline setup.py]: I was planning to add tox. forgot about it. I'll either drop or add one.

File: setup.py (hunk 3 above)
Comment (mindw) [inline setup.py]: https://pypi.python.org/pypi/setuptools_scm/1.7.0 for reference. It uses the tag as the version number and creates a file with the version number. This eliminates a LOT of boiler plate code in the setup.py , conf.py etc.

File: setup.py (hunk 4 above)
Comment (mindw) [inline setup.py]: yes.
(unless you're using features introduced in 3.4)

File: setup.py (hunk 3 above)
Comment (esc) [inline setup.py]: I presume it also works for `build_ext -i`?

**Comment by andreabedini** [comment]
stepping in to say thanks @mindw, I am going to steal all this for my projects :)

File: setup.py [hunk 7]
```diff
@@ -219,29 +126,45 @@ def check_import(pkgname, pkgver):
...
+        'cython>=0.22', 
+        'numpy>=1.7', 
+        'setuptools-scm>1.5.4'
+    ],
+    tests_require=tests_require,
+    extras_require=dict(
+        optional=[
+            'numexpr>=1.4.1', 
+            'pandas', 
+            'pytables'
```
Comment (FrancescAlted) [inline setup.py]: The PyTables package in PyPI is called just 'tables' (https://pypi.python.org/pypi/tables).

BTW, great contribution.  Thanks!

Comment (mindw) [inline setup.py]: yes. will do.

File: setup.py (hunk 3 above)
Comment (mindw) [inline setup.py]: Yes - both Travis and Appveyor passed.

File: RELEASING.rst (hunk 1 above)
Comment (mindw) [inline RELEASING.rst]: In addition to the warning at line 56?

File: setup.py (hunk 6 above)
Comment (mindw) [inline setup.py]: AFAIK, both Cython and a C python extension require a compiler and the Python headers to build.
I'll look into this.

**Comment by mindw** [comment]
Should be better now.

File: setup.py [hunk 8]
```diff
@@ -8,133 +8,35 @@
...
-    else:
-        print_warning(
-            "Numexpr %s installed, but version is not >= %s.  "
-            "Disabling support for it." % (
-                numexpr.__version__, min_numexpr_version))
 
-########### End of checks ##########
+class BuildExtNumpyInc(build_ext):
+    def build_extensions(self):
+        from numpy.distutils.misc_util import get_numpy_include_dirs
```
Comment (andreabedini) [inline setup.py]: if you instead do
```python
numpy_incl = pkg_resources.resource_filename('numpy', 'core/include')
```
you can avoid importing numpy at all, and you can remove `__builtins__.__NUMPY_SETUP__ = False`.

pkg_resources is part of setuptools.

Comment (mindw) [inline setup.py]: While tempting, the above relies on the current Numpy internal structure which is not guaranteed to retain backwards compatibility. The Numpy manual suggests to use get_numpy_include_dirs  ...

Comment (andreabedini) [inline setup.py]: Fair enough.

**Comment by FrancescAlted** [comment]
Everything looks good to me.  +1 for merging this.

**Comment by esc** [comment]
Planning to finish review and merge on friday.

**Comment by esc** [comment]
I won't have time to finish this, please feel free to take over.

**Comment by FrancescAlted** [comment]
Merged.  Still missing an entry in release notes which should follow shortly.

**Comment by FrancescAlted** [comment]
Updated release notes in rev 7222811

**Comment by mindw** [comment]
Please reference this pull request for any issues or questions related to it so I can assist in resolving them.

**Comment by FrancescAlted** [comment]
Sure thing.  Thank you!

**Pull request merged**